Neos.Flow/Classes/Reflection/ReflectionService.php

```diff
@@ -1873,219 +1901,75 @@ protected function loadClassReflectionCompiletimeCache(): bool
-            return (isset($this->availableClassNames[$packageKey]) && in_array($className, $this->availableClassNames[$packageKey], true));
-        };
-
-        foreach ($reflectionData['classesByMethodAnnotations'] as $annotationClassName => $classNames) {
-            $reflectionData['classesByMethodAnnotations'][$annotationClassName] = array_filter($classNames, $methodAnnotationsFilters);
-        }
-
-        $precompiledReflectionStoragePath = $this->getPrecompiledReflectionStoragePath();
-        if (!is_dir($precompiledReflectionStoragePath)) {
-            Files::createDirectoryRecursively($precompiledReflectionStoragePath);
-        }
-        $pathAndFilename = $precompiledReflectionStoragePath . $packageKey . '.dat';
-        file_put_contents($pathAndFilename, extension_loaded('igbinary') ? igbinary_serialize($reflectionData) : serialize($reflectionData));
-    }
-
-    /**
-     * Filter an array of entries where keys are class names by being in the given package namespace.
-     *
-     * @param int|string $packageKey
-     *
-     * @psalm-param array-key $packageKey
-     */
-    protected function filterArrayByClassesInPackageNamespace(array $array, $packageKey): array
-    {
-        return array_filter($array, function ($className) use ($packageKey) {
-            return (isset($this->availableClassNames[$packageKey]) && in_array($className, $this->availableClassNames[$packageKey], true));
-        }, ARRAY_FILTER_USE_KEY);
-    }
-
-    /**
-     * Removes the precompiled reflection data of a frozen package
-     *
-     * This method is used by the package manager.
-     *
-     * @throws FilesException
-     * @throws \Neos\Flow\Utility\Exception
-     */
-    public function unfreezePackageReflection(string $packageKey): void
-    {
-        if (!$this->initialized) {
-            $this->initialize();
-        }
-        $pathAndFilename = $this->getPrecompiledReflectionStoragePath() . $packageKey . '.dat';
-        if (file_exists($pathAndFilename)) {
-            unlink($pathAndFilename);
-        }
-    }
-
     /**
      * Exports the internal reflection data into the ReflectionData cache
      *
      * This method is triggered by a signal which is connected to the bootstrap's
      * shutdown sequence.
      *
-     * If the reflection data has previously been loaded from the runtime cache,
-     * saving it is omitted as changes are not expected.
-     *
-     * In Production context the whole cache is written at once and then frozen in
-     * order to be consistent. Frozen cache data in Development is only produced for
-     * classes contained in frozen packages.
      * @throws Exception
      */
     public function saveToCache(): void
     {
-        if ($this->hasFrozenCacheInProduction()) {
+        if (empty($this->updatedReflectionData)) {
             return;
         }
+
         if (!$this->initialized) {
             $this->initialize();
```

Comment:
initialising here when shutting down doesnt make sense and can never be the case i think - but was so beforehand as well ...